File: go/test/history_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/location-sharing-sdk/go"
	"github.com/voxgig-sdk/location-sharing-sdk/go/core"

	vs "github.com/voxgig-sdk/location-sharing-sdk/go/utility/struct"
)

func TestHistoryEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.History(nil)
		if ent == nil {
			t.Fatal("expected non-nil HistoryEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := historyBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "history." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set LOCATIONSHARING_TEST_HISTORY_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		historyRef01Ent := client.History(nil)
		historyRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "history"}, setup.data), "history_ref01"))

		historyRef01DataResult, err := historyRef01Ent.Create(historyRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		historyRef01Data = core.ToMapAny(historyRef01DataResult)
		if historyRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if historyRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		historyRef01Match := map[string]any{}

		historyRef01ListResult, err := historyRef01Ent.List(historyRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		historyRef01List, historyRef01ListOk := historyRef01ListResult.([]any)
		if !historyRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", historyRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(historyRef01List), map[string]any{"id": historyRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// REMOVE
		historyRef01MatchRm0 := map[string]any{
			"id": historyRef01Data["id"],
		}
		_, err = historyRef01Ent.Remove(historyRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		historyRef01MatchRt0 := map[string]any{}

		historyRef01ListRt0Result, err := historyRef01Ent.List(historyRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		historyRef01ListRt0, historyRef01ListRt0Ok := historyRef01ListRt0Result.([]any)
		if !historyRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", historyRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(historyRef01ListRt0), map[string]any{"id": historyRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func historyBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "history", "HistoryTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read history test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse history test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"history01", "history02", "history03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("LOCATIONSHARING_TEST_HISTORY_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"LOCATIONSHARING_TEST_HISTORY_ENTID": idmap,
		"LOCATIONSHARING_TEST_LIVE":      "FALSE",
		"LOCATIONSHARING_TEST_EXPLAIN":   "FALSE",
		"LOCATIONSHARING_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["LOCATIONSHARING_TEST_HISTORY_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["LOCATIONSHARING_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["LOCATIONSHARING_APIKEY"],
			},
			extra,
		})
		client = sdk.NewLocationSharingSDK(core.ToMapAny(mergedOpts))
	}

	live := env["LOCATIONSHARING_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["LOCATIONSHARING_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/HistoryEntityTest.php
<?php
declare(strict_types=1);

// History entity test

require_once __DIR__ . '/../locationsharing_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class HistoryEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = LocationSharingSDK::test(null, null);
        $ent = $testsdk->History(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = history_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "list", "remove"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "history." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set LOCATIONSHARING_TEST_HISTORY_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $history_ref01_ent = $client->History(null);
        $history_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.history"), "history_ref01"));

        [$history_ref01_data_result, $err] = $history_ref01_ent->create($history_ref01_data, null);
        $this->assertNull($err);
        $history_ref01_data = Helpers::to_map($history_ref01_data_result);
        $this->assertNotNull($history_ref01_data);
        $this->assertNotNull($history_ref01_data["id"]);

        // LIST
        $history_ref01_match = [];

        [$history_ref01_list_result, $err] = $history_ref01_ent->list($history_ref01_match, null);
        $this->assertNull($err);
        $this->assertIsArray($history_ref01_list_result);

        $found_item = sdk_select(
            Runner::entity_list_to_data($history_ref01_list_result),
            ["id" => $history_ref01_data["id"]]);
        $this->assertNotEmpty($found_item);

        // REMOVE
        $history_ref01_match_rm0 = [
            "id" => $history_ref01_data["id"],
        ];
        [$_, $err] = $history_ref01_ent->remove($history_ref01_match_rm0, null);
        $this->assertNull($err);

        // LIST
        $history_ref01_match_rt0 = [];

        [$history_ref01_list_rt0_result, $err] = $history_ref01_ent->list($history_ref01_match_rt0, null);
        $this->assertNull($err);
        $this->assertIsArray($history_ref01_list_rt0_result);

        $not_found_item = sdk_select(
            Runner::entity_list_to_data($history_ref01_list_rt0_result),
            ["id" => $history_ref01_data["id"]]);
        $this->assertEmpty($not_found_item);

    }
}

function history_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/history/HistoryTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = LocationSharingSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["history01", "history02", "history03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("LOCATIONSHARING_TEST_HISTORY_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "LOCATIONSHARING_TEST_HISTORY_ENTID" => $idmap,
        "LOCATIONSHARING_TEST_LIVE" => "FALSE",
        "LOCATIONSHARING_TEST_EXPLAIN" => "FALSE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["LOCATIONSHARING_TEST_HISTORY_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["LOCATIONSHARING_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
            ],
            $extra ?? [],
        ]);
        $client = new LocationSharingSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["LOCATIONSHARING_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["LOCATIONSHARING_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
